explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,260 @@
+```
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.
+
+Note that the only thing that differs between the Intel and AMD variants is the PMU event to be monitored.  Also, note how we are using `sudo` to invoke `perf record`.  This is necessary in order to obtain full access to Linux kernel symbols and relocation information.  It also means that the `openssl-nonpgo.perfdata` file ownership will need to be adjusted:
+
+```bash
+$ sudo chown ubuntu:ubuntu openssl-nonpgo.perfdata
+```
+
+Now, we need to convert the file to `gcov`.
+
+#### Alternative scenario: profiling the program using other approaches
+
+We do not always have well behaved programs that can be profiled like `openssl speed`.  Sometimes, the program might take a long time to finish running the workload, which ends up generating huge `perfdata` files that can prove very challenging for `autofdo` to process.  For example, we profiled QEMU internally and, when executing a workload that took around 3 hours to complete, the `perfdata` generated had a size of several gigabytes.
+
+In order to reduce the size of the `perfdata` that is collected, you might want to play with `perf record`'s `-F` option, which specifies the frequency at which the profiling will be done.  Using `-F 99` (i.e., profile at 99 hertz) is indicated by some `perf` experts because it avoids accidentally sampling the program in lockstep with some other periodic activity.
+
+Another useful trick is to profile in batches instead of running `perf record` for the entire duration of your program's workload.  To do that, you should use the `-p` option to specify a PID to `perf record` while also specifying `sleep` as the program to be executed during the profiling.  For example, this would be the command like we would use if we were to profile a program whose PID is `1234` for 2 minutes:
+
+```bash
+$ sudo perf record \
+	-e br_inst_retired.near_taken \
+	--branch-any \
+	--all-cpus \
+	-p 1234 \
+	--output myprogram.perfdata
+	-- sleep 2m
+```
+
+The idea is that you would run the command above several times while the program being profiled is still running (ideally with an interval between the invocations, and taking care not to overwrite the `perfdata`).  After obtaining all the `perfdata` files, you would then convert them to `gcov` (see section below) and use `profile_merger` to merge all `gcov` files into one.
+
+### Converting `perfdata` to `gcov` with `autofdo`
+
+With `autofdo` installed, you can convert the `perfdata` generated in the last step to `gcov` by doing:
+
+```bash
+
+$ create_gcov \
+	--binary /usr/bin/openssl \
+	--gcov openssl-nonpgo.gcov \
+	--profile openssl-nonpgo.perfdata \
+	--gcov_version 2 \
+	-use_lbr false
+```
+
+`create_gcov` is verbose and will display several messages that may look like something is wrong.  For example, the following output is actually from a successful run of the command:
+
```

Comment:
I know you want to show that even a successful run can throw these, but that is too much.
IMHO the first 6 lines, then ... and then the last 5 lines should be much better.
Otherwise this doesn't even fit on a full-hd screen.

So a) reduce lines please

But also reduce width as this need to scroll a lot left and right.
I think you could strip "third_party/perf_data_converter/src/" from the output without taking away anything of the important message you are telling. Some scrolling will remain, but people can see the important bit right away

Reply:
below you have used "[...] many more lines [...]" - the same works here to stay consistent

Reply:
Done.